pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/ComplexTypeTransformer.java

```diff
@@ -0,0 +1,229 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.recordtransformer;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.ingestion.ComplexTypeConfig;
+import org.apache.pinot.spi.data.readers.GenericRow;
+
+
+/**
+ * A transformer to handle the complex types such as Map and Collection, with flattening and unnesting.
+ * <p>
+ * The map flattening rule will recursively flatten all the map types, except for those under the collection that is not marked as to unnest.
+ *
+ * For example:
+ * <pre>
+ * {
+ *    "t1":{
+ *       "array":[
+ *          {
+ *             "t2":{
+ *                "a":"v1"
+ *             }
+ *          }
+ *       ]
+ *    }
+ * }
+ * </pre>
+ *
+ * flattens to
+ * <pre>
+ * {
+ *    "t1.array":[
+ *       {
+ *          "t2.a":"v1"
+ *       }
+ *    ]
+ * }
+ * <pre/>
+ *
+ * <p>
+ *
+ * The unnesting rule will flatten all the collections provided, which are the paths navigating to the collections. For
+ * the same example above. If the the collectionToUnnest is provided as "t1.array", then the rule will unnest the
+ * previous output to:
+ *
+ * <pre>
+ *  [{
+ *     "t1.arrayt2.a": "v1",
+ *  }]
+ *  * <pre/>
+ *
+ *  Note the unnest rule will output a collection of generic rows under the field {@link GenericRow#MULTIPLE_RECORDS_KEY}.
+ *
+ */
+public class ComplexTypeTransformer implements RecordTransformer {
+  private static final CharSequence DELIMITER = ".";
```

Comment:
If this is turned into a configurable parameter, then local deployments could use any delimiter of their choice, but it seems like the default DELIMITER should be something other than a <DOT> to avoid any possibility of confusion with SQL dot notation.

Reply:
Sounds good. I can have another PR to make this configurable.